### ISSUE TITLE
Cutlass for Attention Decoding

### ISSUE THREAD

**Issue by wimh966** [facebookresearch/xformers#929]
# ❓ Questions and Help
Hi, I want to test the latency of different optimizations in attention decoding phase. Thus, I ran the benchmark_attn_decoding.py file provided by you. I additionally added the case of cutlass like the following in the file.
```
class AttentionDecodingCutlass(AttentionDecodingFlashDecoding):
    OP = xops.fmha.cutlass.FwOp
```
I obtained the latency like this:

![image](https://github.com/facebookresearch/xformers/assets/52482918/41ace7b0-f286-44df-a72d-a31c1388a144)

I find that the latency of cutlass can be much faster than the pytorch version, when the sequence length is small. However, it will be extremely large, and even much larger than that of pytorch version with a long sequence. May I kindly ask is this phenomenon usual? And may I inquire the reason for this? 

I'm using the A100 80G device. Thank you for your help in advance.

**Comment by danthe3rd** [comment]
Hi @wimh966 
This is because the `cutlass` implementation has the pitfalls from `flash-attention v2` that we describe in the Flash-Decoding [blogpost](https://pytorch.org/blog/flash-decoding/).
Mainly, it does not parallelize across the keys/values dimensions, and does not fully utilize the GPU.

**Comment by wimh966** [comment]
Thank you for your reply. I understand. But for the generation phase, I also want to compare the flash decoding with some advanced baselines while the PyTorch implementation is of course very slow. In this way, I can better understand how much benefit removing writing out and loading in the attention weights can bring in the generation phase. Do you have any recommendations? For the context phase, I compare flash attention with cutlass and obtain roughly 2 times of speedup.

Meanwhile, I have another question about GQA (e.g., 8 kv heads). I find that for the context phase, using fewer kv heads would not bring help for latency, but for generation it does. May I kindly ask if is this because we usually allocate each head into each block for context but we can allocate multiple heads into one block for generation as the query vector in the generation phase is very small and we only need to update the attention mask?

Hope that you do not mind if I ask a bit more or maybe some stupid questions. I'm totally a beginner at this and I really appreciate your help. Thanks a lot in advance.

**Comment by danthe3rd** [comment]
> I can better understand how much benefit removing writing out and loading in the attention weights can bring in the generation phase

To fully utilize the GPU, the parallelism must be a multiple of `num_sm * kernel_occupancy`. Assuming you have an A100 GPU, `num_sm=108`. The kernel occupancy for the FW pass will be an integer between 1 and 4.
So you can try to aim for a parallelism of `108*12=1296`.
So I would try with a batch size of 1296, and compare the different approaches with this setting.

> I find that for the context phase, using fewer kv heads would not bring help for latency, but for generation it does

The matrix multiplications are generally not efficient when the matrices are small. That's why we try to increase the batch size etc... During prefiling (or "context phase"), the Q/K and V matrices are pretty big so we don't have this issue.
During generation, K/V are still fairly big, but Q contains only one element. When using GQA, we can permute the sequence length and  head dimensions, which means we can do matrix multiplications on a Q tensor of length 8 (if 8 kv heads) rather than 1. This uses the GPU more efficiently - and this is why you get a speed up.

I hope this helps :)

**Comment by wimh966** [comment]
This definitely helps! Thank you for your detailed reply. For the first question, I understand it works in this way. But I mean do we have any other advanced baselines? For example, flash decoding optimizes both the parallelism and memory transfer. Which operator should I choose if I mainly want to test the benefit brought by memory transfer? I know there is pytorch implementation which utilizes the GPU well, but pytorch implementation is quite slow.

**Comment by danthe3rd** [comment]
All the implementations in xFormers try to reduce memory transfer.
The approaches that increase parallelism (eg Flash-Decoding) have the trade-off of making more memory transfer to allow a higher parallelism.

**Comment by wimh966** [comment]
Hi, I see. But can't we do kv parallelism for cutlass implementation for the generation phase? That will maximize the GPU utilization when sequence length grows longer and batch size grows smaller.

Meanwhile, when I test the latency for cutlass with GQA with 8 groups, I get a bit strange results as below.
|  Units: ms                                           |(256,256,128,16)    |(128,512,128,16)    |(64,1024,128,16)    |(32,2048,128,16)    |(16,4096,128,16)    |(8,8192,128,16)     |(4,16384,128,16)    |(2,32768,128,16)    |(1,65536,128,16)    |(1,131072,128,16)   |(256,256,128,40)    |(128,512,128,40)    |(64,1024,128,40)    |(32,2048,128,40)    |(16,4096,128,40)    |(8,8192,128,40)     |(4,16384,128,40)    |(2,32768,128,40)    |(1,65536,128,40)    |(1,131072,128,40)   |
|------------------------------------------------------|--------------------|--------------------|--------------------|--------------------|--------------------|--------------------|--------------------|--------------------|--------------------|--------------------|--------------------|--------------------|--------------------|--------------------|--------------------|--------------------|--------------------|--------------------|--------------------|--------------------|
| pytorch                                              |2.67                |2.37                |2.37                |2.38                |2.38                |2.40                |2.44                |2.44                |0.54                |1.20                |5.97                |6.23                |6.27                |6.27                |6.29                |6.28                |6.31                |6.32                |0.97                |1.92                |
| cutllass                                             |0.39                |0.38                |0.38                |0.38                |0.45                |0.52                |0.69                |1.30                |2.58                |5.14                |0.94                |0.91                |0.89                |0.88                |0.88                |0.92                |1.06                |1.40                |2.59                |5.16                |
| flash                                                |0.39                |0.37                |0.37                |0.37                |0.43                |0.52                |0.33                |0.36                |0.36                |0.68                |0.99                |0.94                |0.91                |0.90                |0.89                |0.91                |1.06                |0.81                |0.79                |1.55                |
| pytorch_gqa                                          |2.31                |2.39                |2.43                |2.42                |2.38                |2.40                |2.44                |2.44                |0.54                |1.19                |6.09                |6.31                |6.35                |6.33                |6.36                |6.33                |6.34                |6.32                |0.97                |1.92                |
| cutlass_gqa                                         |0.87                |0.86                |0.85                |0.86                |0.86                |0.86                |1.32                |2.59                |5.13                |10.21               |0.85                |0.86                |0.86                |0.86                |0.90                |0.88                |1.32                |2.59                |5.12                |10.20               |
| flash_gqa                                            |0.20                |0.20                |0.20                |0.23                |0.27                |0.18                |0.19                |0.19                |0.20                |0.36                |0.20                |0.20                |0.20                |0.23                |0.27                |0.18                |0.19                |0.19                |0.20                |0.36                |

Column name indicates (batch size, seq_len, head_dim, head_num) and the row with _gqa suffix indicates taking 8 kv heads otherswise #head_num kv heads. 
You can find that pytorch implementation shows similar results with and without GQA. Flash implementation shows better results with GQA. I think these are normal phenomenons. However, cutlass_gqa is roughly 2 times of cutlass, which I think is abnormal. Do you have any idea for this?

**Comment by wimh966** [comment]
Decoding. I test the generation phase and Q 1. Thank you.

**Comment by danthe3rd** [comment]
Then with cutlass for GQA (say 4 query heads per kv head), cutlass will launch 8x less blocks, and will most likely not use the entire GPU.

Instead of launching 4 blocks for 4 query-heads to do something like this:
<img width="610" alt="image" src="https://github.com/facebookresearch/xformers/assets/43445237/1008e626-4dad-4db9-8bc6-5c4ba6ccec74">

It does something like that:
<img width="304" alt="image" src="https://github.com/facebookresearch/xformers/assets/43445237/8ed4f6f4-27d8-473d-97cd-e8f129512e9b">

This uses less compute, but is also less parallel. If you increase the batch size dramatically (until cutlass can fill the GPU), you will see that GQA is faster than without

**Comment by wimh966** [comment]
I see. So it's because that cutlass implementation does not apply kv parallelism in this repo that it runs slowly with long sequences and even much slower with GQA. 

Thank you so much!!!

**Comment by danthe3rd** [comment]
Exactly :)
